convert.go

```diff
@@ -82,10 +83,103 @@ func (c *conversion) putBuffer(b *conversionBuffer) {
 	c.buffers.Put(b)
 }
 
+// convertFunc takes a target and source row, then copies data
+// from the source to the target according to the heuristics determined
+// by the target schema given in makeConvertFunc
+type convertFunc func(Row, *conversionBuffer) (Row, error)
+
+func (c *conversion) makeConvertFunc(node Node) (convert convertFunc) {
+	if !node.Leaf() {
+		_, convert = c.convertFuncOf(0, node)
+	}
+	return convert
+}
+
+func (c *conversion) convertFuncOf(tgtIdx int16, node Node) (int16, convertFunc) {
+	switch {
+	case node.Repeated():
+		return c.convertFuncOfRepeated(tgtIdx, node)
+	default:
+		return c.convertFuncOfRequired(tgtIdx, node)
+	}
+}
+
+func (c *conversion) convertFuncOfRequired(tgtIdx int16, node Node) (int16, convertFunc) {
+	switch {
+	case node.Leaf():
+		return c.convertFuncOfLeaf(tgtIdx, node)
+	default:
+		return c.convertFuncOfGroup(tgtIdx, node)
+	}
+}
+
+// convertFuncOfLeaf is the base case to our schema-tree traversal, doing the
+// actual copy of the value from the old Row (via conversionBuffer) to the new Row
+func (c *conversion) convertFuncOfLeaf(tgtIdx int16, node Node) (int16, convertFunc) {
+	return tgtIdx + 1, func(tgt Row, src *conversionBuffer) (Row, error) {
+		value := Value{}
+		if tgtIdx >= 0 && len(src.columns[tgtIdx]) > 0 {
+			// Pop the top value and remove from the buffer.
+			value = src.columns[tgtIdx][0]
+			src.columns[tgtIdx] = src.columns[tgtIdx][1:]
+		}
+		value.kind = ^int8(c.targetColumnKinds[tgtIdx])
+		value.columnIndex = ^tgtIdx
+		tgt = append(tgt, value)
+		return tgt, nil
+	}
+}
+
+func (c *conversion) convertFuncOfGroup(tgtIdx int16, node Node) (int16, convertFunc) {
+	fields := node.Fields()
+	funcs := make([]convertFunc, len(fields))
+
+	for i, field := range fields {
+		tgtIdx, funcs[i] = c.convertFuncOf(tgtIdx, field)
+	}
+
+	return tgtIdx, func(tgt Row, src *conversionBuffer) (Row, error) {
+		var err error
+		for i, convFunc := range funcs {
+			if tgt, err = convFunc(tgt, src); err != nil {
+				err = fmt.Errorf("%s → %w", fields[i].Name(), err)
+				break
+			}
+		}
+		return tgt, err
+	}
+}
+
+func (c *conversion) convertFuncOfRepeated(tgtIdx int16, node Node) (int16, convertFunc) {
+	nextIdx, convFunc := c.convertFuncOf(tgtIdx, Required(node))
+	return nextIdx, func(tgt Row, src *conversionBuffer) (Row, error) {
+		var err error
+
+		for i, elem := range src.columns[tgtIdx] {
+			// if our repetitionLevel is not equal to the definition level on our
+			// second or subsequent copies, then this element belongs to another
+			// repeated object above us.
+			if i != 0 && int16(elem.repetitionLevel) != int16(elem.definitionLevel) {
+				break
```

Comment:
I'm pretty sure this logic holds 100% of the time, but would appreciate a second opinion here!

Reply:
Maybe a counter example would be a case of this column structure:
```
message M {
  repeated group A {
    optional int64 B;
  }
}
```
In which case values of the column A.B may have a definition level that differs from the repetition level?
- a null value in A.B would have definition level 1
- a non-null value in A.B would have definition level 2
- the repetition level is 0 on the value of the first record, and 2 on the other values

Reply:
I'll see if I can't produce a failing test case for this scenario, or otherwise come up with a good explanation of the behavior in this case.

Reply:
Got a failing test case, we'll see what I can do with this.  Might just have to pass in the levels object, we'll see!

Reply:
Keeping track of the expected repetition level is probably gonna be an effective strategy (this is what we do in general across the library).

Reply:
Seems like there's no way around it from what I can tell